src/app/file-manager/login/login.component.spec.ts

```diff
@@ -1,45 +1,115 @@
+import { Router } from '@angular/router';
+import { Location } from '@angular/common';
+
+class RouterStub {
+  navigate(URL: any[]) {
+    return URL;
+  }
+}
 
 describe('LoginComponent', () => {
   let component: LoginComponent;
-  let fixture: ComponentFixture<LoginComponent>;
+  let loginComponent: ComponentFixture<LoginComponent>;
+  let electronIpc: ElectronIpcService;
+  let loginService: LoginService;
+  let routerMod: Router;
+  let location: Location;
 
   beforeEach(async(() => {
     TestBed.configureTestingModule({
       declarations: [ LoginComponent ],
       imports: [
         BrowserAnimationsModule,
-        RouterModule,
         // core & shared
         CoreModule,
         SharedModule,
 
         // features
         SettingsModule,
-        NgxElectronModule,
-        RouterTestingModule,
+        RouterTestingModule
       ],
       providers: [
-         FirecloudService,
+        FirecloudService,
+        LoginService,
+        ElectronIpcService,
+        RouterStub
       ]
-    })
-    .compileComponents();
+    });
+
+    TestBed.overrideComponent(LoginComponent, {
+      set: {
+        providers: [
+          {provide: FirecloudService, useClass: FirecloudApiMockService},
+          {provide: LoginService, useClass: LoginMockService},
+          {provide: ElectronIpcService, useClass: ElectronIpcMockService},
+          {provide: Router, useClass: RouterStub}
+        ]
+      }
+    });
+
+    loginComponent = TestBed.createComponent(LoginComponent);
+    component = loginComponent.componentInstance;
+    loginComponent.detectChanges();
+    // electronIpc should be the mocked class
+    electronIpc = loginComponent.debugElement.injector.get(ElectronIpcService);
+    loginService = loginComponent.debugElement.injector.get(LoginService);
+    routerMod = loginComponent.debugElement.injector.get(Router);
+
+    location = TestBed.get(Location);
   }));
 
-  beforeEach(() => {
-    fixture = TestBed.createComponent(LoginComponent);
-    component = fixture.componentInstance;
-    fixture.detectChanges();
+  describe('Create LoginComponent', () => {
+
+    it('should create the LoginComponent', () => {
+      expect(electronIpc instanceof ElectronIpcMockService).toBeTruthy();
+      expect(loginService instanceof LoginMockService).toBeTruthy();
+      expect(routerMod instanceof RouterStub).toBeTruthy();
+      expect(component).toBeTruthy();
+    });
+
+    it('should display the login button', () => {
+      loginComponent.detectChanges();
+      const compiled = loginComponent.debugElement.nativeElement;
+      expect(compiled.querySelector('button').textContent).toContain('Sign In');
+    });
+  });
+
+  describe('Log the user', () => {
+    it('should log', () => {
+      component.logIn();
+      expect(component.hasError).toBeFalsy();
+    });
+  });
+
+  describe('Show errors', () => {
+    it('should show messages for 401 error code', () => {
+      component.showError(401);
```

Comment:
This is OK but it would be better to drive the error from the service layer, e.g. have the login request return the http status code and then verify that the error show in the view is correct.

Reply:
Sure, I'm looking how to implement that.

Reply:
This is taking me too much time to solve. I'll tackle this in the next approach with the following tests.